semantics/library/bootstrap.k

```diff
@@ -49,14 +58,50 @@ module LIBC-BOOTSTRAP
 
      syntax List ::= makeVarArgs(List) [function]
                    | makeVarArgsAux(List) [function]
-     rule makeVarArgsAux(L:List ListItem(X:CId) ListItem(vararg(0)))
-          => L ListItem(X) ListItem(varg(incSymbolic(&(X))))
-     rule makeVarArgsAux(L:List ListItem(X:CId))
-          => makeVarArgsAux(L)
-          requires X =/=K vararg(0)
+
      rule makeVarArgs(L:List) => makeVarArgsAux(L)
           requires vararg(0) in L
      rule makeVarArgs(L:List ListItem(X:CId))
           => L ListItem(X) ListItem(varg(incSymbolic(&(X))))
           requires notBool vararg(0) in (L ListItem(X))
+
+     rule makeVarArgsAux(L:List ListItem(X:CId) ListItem(vararg(0)))
+          => L ListItem(X) ListItem(varg(incSymbolic(&(X))))
+     rule makeVarArgsAux(L:List ListItem(X:CId))
+          => makeVarArgsAux(L)
+          requires X =/=K vararg(0)
+
+     rule prepareBuiltin(Identifier(F:String), .HeatList)
+          => builtin(F)
+          [structural]
+     rule prepareBuiltin(Identifier(F:String),
+               hrItem(A1:RValue))
+          => builtin(F, A1)
+          [structural]
+     rule prepareBuiltin(Identifier(F:String),
+               hrItem(A1:RValue) hs:: hrItem(A2:RValue))
+          => builtin(F, A1, A2)
+          [structural]
+     rule prepareBuiltin(Identifier(F:String),
+               hrItem(A1:RValue) hs:: hrItem(A2:RValue) hs:: hrItem(A3:RValue))
+          => builtin(F, A1, A2, A3)
+          [structural]
+     rule prepareBuiltin(Identifier(F:String),
+               hrItem(A1:RValue) hs:: hrItem(A2:RValue) hs:: hrItem(A3:RValue) hs:: hrItem(A4:RValue))
+          => builtin(F, A1, A2, A3, A4)
+          [structural]
+     rule prepareBuiltin(Identifier(F:String),
+               hrItem(A1:RValue) hs:: hrItem(A2:RValue) hs:: hrItem(A3:RValue) hs:: hrItem(A4:RValue) hs:: hrItem(A5:RValue))
+          => builtin(F, A1, A2, A3, A4, A5)
+          [structural]
+     rule prepareBuiltin(Identifier(F:String),
+               hrItem(A1:RValue) hs:: hrItem(A2:RValue) hs:: hrItem(A3:RValue)
+               hs:: hrItem(A4:RValue) hs:: hrItem(A5:RValue) hs:: hrItem(A6:RValue))
+          => builtin(F, A1, A2, A3, A4, A5, A6)
+          [structural]
+     rule prepareBuiltin(Identifier(F:String), Args:HeatList)
+          => builtin*(F, hListToList(Args))
```

Comment:
I think you're going to see a confluence error here because there is a critical pair between this rule and the above rules. You probably want to assert something about the length of the list in the side condition here...